scripts/benchmark.sh

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env bash
+# Benchmark pipes.sh by revisions
+# Copyright (c) 2018 Pipeseroni/pipes.sh contributors
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# Usage: scripts/benchmark.sh <revision>...
+#
+# Example:
+#
+#     $ scripts/benchmark.sh f7d0941 $(git tag) HEAD
+#     f7d0941   :    564 c/s
+#     2013-02-01:    534 c/s
+#     2013-02-06:    534 c/s
+#     [snip]
+#     v1.2.0    :    529 c/s
+#     v1.3.0    :    492 c/s
+#     HEAD      :   3030 c/s
+
+
+PIPESSH='pipes.sh'
+LIMIT=1000
+
+SED_E=(
+    -e 's/read\|sleep\|cat/:/g'  # NOP
+    -e 's/! :/:/'                # ! read -> ! : -> :
+)
+SED_CLNUP=("${SED_E[@]}" -e 's/&& t=0/\&\& cleanup/')
+SED_BREAK=("${SED_E[@]}" -e 's/&& t=0/\&\& break/')
+
+
+# monkey-patch gitrevision $1 pipes.sh:
+# 1. make read/sleep/cat nop
+# 2. exit when reach -r LIMIT
+#
+# Some early commits use sleep to work around Bash < 4 for delay < 1 second,
+# and cat is used to clear out standard input, which could hang if not made
+# nop.
+mkp_pipes() {
+    local GIT=(git cat-file --textconv "$1:$PIPESSH")
+    if "${GIT[@]}" | grep cleanup &>/dev/null; then
+        "${GIT[@]}" | sed "${SED_CLNUP[@]}"
+    else
+        "${GIT[@]}" | sed "${SED_BREAK[@]}"
+    fi
+}
+
+
+# time the monkey-patched gitrevision $1 pipes.sh with Bash builtiin time.
+# The stdout and stderr are swapped.  So the pipes are on stderr and time's
+# result (real %e) is on stdout.
+run() {
+    TIME=%e time bash <(mkp_pipes "$1") -r $LIMIT 3>&2 2>&1 1>&3
```

Comment:
I think this is calling the `time` executable because of the preceding `TIME=%e`. On my system, where I don't have a `time` executable, `TIME=%e time echo 1` says `bash: time: command not found`.

The following worked for me:

```bash
run() {
    # TIMEFORMAT sets format for builtin time
    local TIMEFORMAT=%R
    # Point fd 3 at stderr
    exec 3>&2
    # Redirect child process' stdout to stderr and stderr to stdout.
    { time bash <(mkp_pipes "$1") -r $LIMIT 1>&3; } 2>&1
}
```

Reply:
I was completely in oblivion, even thinking I got it right when I removed `command` from `command time`, because the output was correct.  I was reading `time(1)` and thought Bash's time also use `TIME` as format, RTFWM (Read The Fucking Wrong Manual), or I would have caught my mistake at that moment.  I believed I thought I was reading `help time`, but instead, it's `time(1)`.

Originally, I intended to use `/usr/bin/time`, glad I decided to switch to Bash keyword `time` (I mistook it as a builtin, but it made sense since it actually times a pipeline), so you caught my mistake.  (I have forgot installing `sys-process/time-1.7-r1` *seven and a half years ago* (2009-10-18) and it's still one of three stable versions in Portage tree.)

Reply:
The same thing happens to me every single time I try and use `time` for anything non-trivial. It would have been nice if bash's builtin `time` obeyed the `time(1)` interface, but I suppose we're stuck with it now.